### PR TITLE
AWQ: Up to 2.66x higher throughput

### PR DESCRIPTION
**Pull request by casper-hansen**
The strategy is to dequantize and run FP16 matmul for longer sequences. This could probably be faster if we just used cublas instead of `torch.matmul`.

EDIT: It seems throughput can be over 2x in vLLM because context processing is such a crucial part of the framework.

```
python benchmarks/benchmark_throughput.py --input-len 1024 --output-len 64 --model TheBloke/Mistral-7B-Instruct-v0.2-AWQ --quantization awq --num-prompts 100 --max-model-len 1070 --dtype half
```

Tested on 1x A100 (80GB).

Input Len | Threshold | Main (requests/s, tokens/s) | PR (requests/s, tokens/s) | Speedup (requests/s)
-- | -- | -- | -- | --
16384 | 256 | 0.19, 3138.15 | 0.54, 8879.04 | 2.84x
4096 | 1024 | 0.84, 3500.00 | 2.24, 9336.44 | 2.66x
1024 | 1024 | 3.28, 3570.90 | 6.70, 7286.77 | 2.04x
512 | 512 | 5.96, 3435.14 | 11.91, 6860.50 | 1.99x
256 | 256 | 10.39, 3323.90 | 18.49, 5916.05 | 1.78x

### CODE REVIEW

**Comment by MichaelJayW** [comment]
Does it affect the accuracy?

**Comment by fxmarty** [comment]
@casper-hansen that's really cool, in line with the bench [here](https://github.com/AutoGPTQ/AutoGPTQ/pull/484) where indeed cublas (that exllama kernel is using for longer sequences) is just better than the AWQ GEMM kernel.

I think it would make our life easier if we had the same kind of dispatch for marlin.

**Comment by casper-hansen** [comment]
> Does it affect the accuracy?

This should have no impact on accuracy. The dequantization kernel is strictly equivalent to the dequantization from the GEMM kernel. 

> @casper-hansen that's really cool, in line with the bench [here](https://github.com/AutoGPTQ/AutoGPTQ/pull/484) where indeed cublas (that exllama kernel is using for longer sequences) is just better than the AWQ GEMM kernel.
> 
> I think it would make our life easier if we had the same kind of dispatch for marlin.

Yes, I agree that the Marlin kernels could achieve even higher throughput. The most crucial part is just missing - it’s only for symmetric quantization.

**Comment by fxmarty** [comment]
is it that crucial though? Many int4*fp16 models use symmetric weight quantization successfully

**Comment by casper-hansen** [comment]
> is it that crucial though? Many int4*fp16 models use symmetric weight quantization successfully

It may turn out to just be an engineering problem, but from my limited experience, the most popular symmetric weight quantization methods suffer from a higher quantization error.

File: setup.py [hunk 1]
```diff
@@ -255,7 +255,9 @@ def get_torch_arch_list() -> Set[str]:
 ]
 
 if _is_cuda():
-    vllm_extension_sources.append("csrc/quantization/awq/gemm_kernels.cu")
+    vllm_extension_sources.extend([
```
Comment (WoosukKwon) [inline setup.py:258]: What is the purpose of this change?

File: csrc/quantization/awq/gemm_kernels.cu [hunk 2]
```diff
@@ -493,9 +493,117 @@ __global__ void __launch_bounds__(64) gemm_forward_4bit_cuda_m16n64k32(int G, in
 #endif
 }
 
+__global__ void __launch_bounds__(64) dequantize_weights(
```
Comment (WoosukKwon) [inline csrc/quantization/awq/gemm_kernels.cu:496]: Just wondering: is this also supported by T4?

File: vllm/model_executor/layers/quantization/awq.py [hunk 3]
```diff
@@ -153,7 +153,16 @@ def apply_weights(self,
         pack_factor = self.quant_config.pack_factor
         out_shape = (x.shape[:-1] + (qweight.shape[-1] * pack_factor, ))
         reshaped_x = x.reshape(-1, x.shape[-1])
-        out = ops.awq_gemm(reshaped_x, qweight, scales, qzeros, pack_factor)
+
+        # batch_size*seq_len >= threshold
+        FP16_MATMUL_HEURISTIC_CONDITION = x.shape[0] * x.shape[1] >= 256
```
Comment (WoosukKwon) [inline vllm/model_executor/layers/quantization/awq.py:158]: I think we can make it a bit more general so that 1D tensors can also be supported?
```suggestion
        FP16_MATMUL_HEURISTIC_CONDITION = x.shape[:-1].numel() >= 256
```

**Review by WoosukKwon — COMMENTED** [review]
Hi @casper-hansen, thanks for submitting the PR! Left some minor comments. Please take a look.

File: csrc/quantization/awq/gemm_kernels.cu (hunk 2 above)
Comment (casper-hansen) [reply csrc/quantization/awq/gemm_kernels.cu:496]: Yes it should be. The CC 7.5 support was only for the mma.sync.aligned.m16n8k16 instruction. This should be compatible as it’s already in the GEMM kernel.

File: setup.py (hunk 1 above)
Comment (casper-hansen) [reply setup.py:258]: This is an artifact of an earlier commit in my branch. Will fix

File: vllm/model_executor/layers/quantization/awq.py (hunk 3 above)
Comment (casper-hansen) [reply vllm/model_executor/layers/quantization/awq.py:158]: Yes, this is a bit more general although I’m not familiar enough with 1D shapes occurring in vLLM. Let me see if the suggestion works

File: csrc/quantization/awq/gemm_kernels.cu (hunk 2 above)
Comment (WoosukKwon) [reply csrc/quantization/awq/gemm_kernels.cu:496]: Sounds good! Thanks for the clarification.

File: vllm/model_executor/layers/quantization/awq.py (hunk 3 above)
Comment (casper-hansen) [reply vllm/model_executor/layers/quantization/awq.py]: Fixed!

File: setup.py (hunk 1 above)
Comment (casper-hansen) [reply setup.py]: Fixed!

**Comment by casper-hansen** [comment]
> Hi @casper-hansen, thanks for submitting the PR! Left some minor comments. Please take a look.

@WoosukKwon Thanks for the review. I applied your suggested fixes and tested that throughput is as expected.

**Review by WoosukKwon — APPROVED** [review]
LGTM! thanks for the fix!

**Comment by sitabulaixizawaluduo** [comment]
I tested Llama-13b on A30 with tensor parallel size is 4, and I found awq throughput is lower than fp16.

**Comment by casper-hansen** [comment]
> I tested Llama-13b on A30 with tensor parallel size is 4, and I found awq throughput is lower than fp16.

This is as expected. You cannot exceed W16A16 performance with W4A16 when you test for throughput. You would need W4A4 (Atom, lower quality model) or W8A8 (SmoothQuant, also lower quality model).

This is because W4A16 methods require dequantization, so when you test throughput, you become compute bound and then it limits the performance.

File: vllm/model_executor/layers/quantization/awq.py [hunk 4]
```diff
@@ -153,7 +153,16 @@ def apply_weights(self,
...
         out_shape = (x.shape[:-1] + (qweight.shape[-1] * pack_factor, ))
         reshaped_x = x.reshape(-1, x.shape[-1])
-        out = ops.awq_gemm(reshaped_x, qweight, scales, qzeros, pack_factor)
+
+        # num_tokens >= threshold
+        FP16_MATMUL_HEURISTIC_CONDITION = x.shape[:-1].numel() >= 256
+
+        if FP16_MATMUL_HEURISTIC_CONDITION:
+            out = ops.awq_dequantize(qweight, scales, qzeros, 0, 0, 0)
+            out = torch.matmul(reshaped_x, out)
```
Comment (zcnrex) [inline vllm/model_executor/layers/quantization/awq.py:162]: Curious to learn: would this copy the dequantized weights back to the memory before doing torch.matmul? And a potential optimization is through implementing a more efficient mixed precision matmul that saves 1 data transfer to the memory?

Comment (casper-hansen) [reply vllm/model_executor/layers/quantization/awq.py:162]: You are probably right that there is potential to eliminate overhead. Exllama runs dequantization and then directly calls cublas for matmul inside the same CUDA kernel. Definitely something to explore!
